src/os/bluestore/bluestore_dump.cc

```diff
@@ -0,0 +1,438 @@
+#include "common/admin_socket.h"
+#include "common/url_escape.h"
+namespace po = boost::program_options;
+
+#include "bluestore_dump.h"
+
+void Dump::dump(const bluestore_blob_t& t)
+{
+  f->open_array_section("extents");
+  for(auto& e: t.get_extents()) {
+    f->open_object_section("extent");
+    if ((int64_t)e.offset != -1)
+      f->dump_format_unquoted("o", "0x%x", e.offset);
+    else
+      f->dump_int("o", -1);
+    f->dump_format_unquoted("l", "0x%x", e.length);
+    f->close_section();
+  }
+  f->close_section();
+  f->dump_format_unquoted("flags", "%d(%s)", t.flags, bluestore_blob_t::get_flags_string(t.flags).c_str());
+  if (t.is_compressed()) {
+    f->dump_unsigned("logical_length", t.get_logical_length());
+    f->dump_unsigned("compressed_length", t.get_compressed_payload_length());
+  }
+  if (t.has_csum()) {
+    f->open_object_section("csum");
+    f->dump_format_unquoted("type", "%d(%s)", t.csum_type, Checksummer::get_csum_type_string(t.csum_type));
+    f->dump_int("order", t.csum_chunk_order);
+    char str[t.csum_data.length()*2+1];
+    for (size_t i = 0; i < t.csum_data.length(); i++)
+      sprintf(&str[i*2], "%2.2x", *(unsigned char*)(t.csum_data.c_str()+i));
+    f->dump_format_unquoted("data", "%s", str);
+    f->close_section();
+  }
+  if (t.has_unused()) {
+    f->dump_unsigned("unused", t.unused);
+  }
+}
+
+
+void Dump::dump(bluestore_onode_t& t)
+{
+  f->dump_unsigned("nid", t.nid);
+  f->dump_unsigned("size", t.size);
+  f->open_object_section("attrs");
+  for (auto p = t.attrs.begin(); p != t.attrs.end(); ++p) {
+    f->open_object_section("attr");
+    f->dump_string("name", p->first.c_str());  // it's not quite std::string
+    f->dump_string("val", url_escape(std::string(p->second.c_str(), p->second.length())));
+    f->close_section();
+  }
+  f->close_section();
+  f->dump_format_unquoted("flags", "%d(%s)", t.flags, t.get_flags_string().c_str());
+  f->open_array_section("extent_map_shards");
+  for(auto& e: t.extent_map_shards) {
+    f->open_object_section("shard");
+    f->dump_format_unquoted("offset", "0x%x", e.offset);
+    f->dump_unsigned("bytes", e.bytes);
+    f->close_section();
+  }
+  f->close_section();
+  f->dump_unsigned("expected_object_size", t.expected_object_size);
+  f->dump_unsigned("expected_write_size", t.expected_write_size);
+  f->dump_unsigned("alloc_hint_flags", t.alloc_hint_flags);
+}
+
+void Dump::dump(const bluestore_blob_use_tracker_t& t)
+{
+  f->dump_int("au_size", t.au_size);
+  if (t.au_size) {
+    f->dump_int("num_au", t.num_au);
+    if (!t.num_au) {
+      f->dump_int("total_bytes", t.total_bytes);
+    } else {
+      f->open_array_section("bytes_per_au");
+      for (size_t i = 0; i < t.num_au; ++i) {
+	f->dump_int("bytes", t.bytes_per_au[i]);
+      }
+      f->close_section();
+    }
+  }
+}
+
+
+//equivalent to BlueStore::ExtentMap::decode_spanning_blobs
+int Dump::decode_spanning_blobs(std::map<int16_t, bluestore_blob_t>& spanning_blobs,
```

Comment:
Can we use all such equivalents from BlueStore directly instead? Well, most of such copy-pastes will go away anyway if dump implementations are members of respective classes.